Boolean (checkbox) params not setting correct default on toggle

dlbeswick opened the pull request:
This one is more for discussion at this stage, I just made a pull request so you can easily see the code changes.

I play a lot with Buzztrax in looping mode while I fiddle with parameters, and I've seen a few issues with parameters flipping back to old values (or in this case, values from two changes ago) when the song loops. Generally, the value's "pattern default" will not have been set properly at some point during the UI event. The issue here was seen when toggling a boolean property (checkbox.)

"update_param_after_interaction" is the call that sets the pattern default, according to the current property value. 

"on_checkbox_property_toggled" via the "toggled" signal is the code that actually updates the property on the object, but "on_button_release_event" via "button-release-event" is called first, and it seems that the object's property hasn't been set yet, so the wrong "default" is set for the property.

Making the changes below did fix the problem, but I'm not really sure of the "right" fix.

ensonic commented:
I see. The calls are all there, but the ordering is different.  If this helps, I won't mind merging it. Do we need this for any other type of control or did you only observe this for checkbuttons?

dlbeswick commented:
Hello, so sorry about the delay in replying. I have seen this happen with other controls in other situations, the one that comes to mind is with sliders if you use the mousewheel, or drag the slider while holding the mouse. If you don't release the mouse button and the song loops, then the value resets to the value it was before the mouse move started. I'll fix these as I see them, too.

I think this is definitely needed -- it was mysterious for me loading up Buzztrax for the first time and fiddling with the SID machine in a short loop, but my checkbox settings seemed to keep flicking back and forth.

I know that this PR is a little messy with the commenting and so on, but I figured you may have had a reason for the code change and perhaps this "fix" would upset that. So I didn't want to totally remove the code before we knew for sure. I haven't seen any issues so far in my fork.

dlbeswick commented:
The volume widget on the wires is another one.

dlbeswick commented:
And before I forget, this also happens with the right-click "Reset parameter" menu

dlbeswick commented:
I feel more confident in this fix now, having looked at it again.

I've modified all the range controls, removing the connection to the "release button" signal, which is where setting values used to be done. Instead, I update the default values on any "change" event, so it works while dragging the slider as well as using the keys to adjust values. If we don't do this, then the values will flick back to the defaults when the song loops while the user is dragging a slider.

I also fixed the "Reset/Reset all parameters" menu options.

dlbeswick commented:
Possibly this would still happen when using keys to change combo box selections, but this is probably not a common and could be fixed in future.

ensonic commented:
Thanks!
